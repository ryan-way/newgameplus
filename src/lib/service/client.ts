import type { AppRouter } from '../../electron/router/renderer';
import { ipcLink } from '../../electron/router/renderer';
import { createTRPCClient } from '@trpc/client';

const trpcClient = createTRPCClient<AppRouter>({
  links: [ipcLink()],
});

export default trpcClient;
