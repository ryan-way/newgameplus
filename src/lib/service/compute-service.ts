import trpcClient from './client';

type Token = 'X' | 'O' | ' ';
type Row = [Token, Token, Token];
type Board = [Row, Row, Row];

export class ComputeService {
  public count(count: number): Promise<number> {
    return trpcClient.query('compute-count', count);
  }

  public tictactoe(board: Board) {
    return trpcClient.query('compute-tictactoe', board);
  }
}
