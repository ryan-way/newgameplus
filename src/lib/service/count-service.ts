import { Count } from '../../models';
import trpcClient from './client';

export class CountService {
  public first(): Promise<Count> {
    return trpcClient.query('count-first');
  }

  public read(id: number): Promise<Count> {
    return trpcClient.query('count-read', id);
  }

  public readAll(): Promise<Count[]> {
    return trpcClient.query('count-read-all');
  }

  public update({ id, count }: { id: number; count: number }): Promise<Count> {
    return trpcClient.mutation('count-update', { id: id, count: count });
  }

  public create({ count }: { count: number }): Promise<Count> {
    return trpcClient.mutation('count-update', { count: count });
  }
}
