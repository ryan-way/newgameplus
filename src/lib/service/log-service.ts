import trpcClient from '../../trpc/client';

export class LogService {
  public info(message: string) {
    trpcClient.mutation('log-info', message);
  }

  public error(message: string) {
    trpcClient.mutation('log-error', message);
  }

  public metric(category: string, operation: string, params: string, measurement: number) {
    trpcClient.mutation('log-metric', {
      category: category,
      operation: operation,
      params: params,
      measurement: measurement,
    });
  }
}
