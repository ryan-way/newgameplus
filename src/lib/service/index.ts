import { CountService } from './count-service';
import { ComputeService } from './compute-service';

const countService = new CountService();
const computeService = new ComputeService();

export { countService, computeService };
