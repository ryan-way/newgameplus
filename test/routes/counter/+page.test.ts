import { expect } from 'vitest';
import { render, RenderResult, fireEvent } from '@testing-library/svelte';
import { countService, computeService } from '$lib/service';
import { tick } from 'svelte';
import Counter from '../../../src/routes/counter/+page.svelte';

vi.mock('$lib/service', () => {
  const countService = {
    first: vi.fn(() => Promise.resolve({ id: 1, count: 0 })),
    update: vi.fn(),
  };

  const computeService = {
    count: vi.fn(x => Promise.resolve(x + 1)),
  };
  return {
    countService,
    computeService,
  };
});

describe('Counter', () => {
  let renderedComponent: RenderResult<Counter>;

  describe('when the component is rendered', () => {
    beforeEach(() => {
      countService.first = vi.fn(() => Promise.resolve(null));
      renderedComponent = render(Counter);
    });

    test('count should not be rendered', () => {
      expect(() => renderedComponent.getByTestId('count')).toThrow();
    });

    test('increment button should not be rendered', () => {
      expect(() => renderedComponent.getByTestId('increment')).toThrow();
    });

    test('decrement button should not be rendered', () => {
      expect(() => renderedComponent.getByTestId('decrement')).toThrow();
    });

    test('reset button should not be rendered', () => {
      expect(() => renderedComponent.getByTestId('reset')).toThrow();
    });

    test('should show the proper heading', () => {
      const element = renderedComponent.getByTestId('counter');
      expect(element).not.toBeNull();
      expect(element.innerHTML).toMatchSnapshot();
    });

    test('should match snapshot', () => {
      expect(renderedComponent.innerHTML).toMatchSnapshot();
    });
  });

  describe('after count has returned', () => {
    beforeEach(() => {
      countService.first = vi.fn(() => Promise.resolve({ id: 1, count: 0 }));
      renderedComponent = render(Counter);
    });

    describe('updated component', () => {
      test('should render increment button', () => {
        const element = renderedComponent.getByTestId('increment');
        expect(element).not.toBeNull();
        expect(element.innerHTML).toMatchSnapshot();
      });

      test('should render decrement button', () => {
        const element = renderedComponent.getByTestId('decrement');
        expect(element).not.toBeNull();
        expect(element.innerHTML).toMatchSnapshot();
      });

      test('should render reset button', () => {
        const element = renderedComponent.getByTestId('reset');
        expect(element).not.toBeNull();
        expect(element.innerHTML).toMatchSnapshot();
      });

      test('should render counter as zero', () => {
        const element = renderedComponent.getByTestId('count');
        expect(element).not.toBeNull();
        expect(element.innerHTML).toMatchSnapshot();
      });

      test('should match snapshot', () => {
        expect(renderedComponent.innerHTML).toMatchSnapshot();
      });

      describe('when a button is clicked', () => {
        test('increment should render 1', async () => {
          const increment = renderedComponent.getByTestId('increment');

          await fireEvent.click(increment);
          await tick();

          const count = renderedComponent.getByTestId('count');
          expect(count.innerHTML).toBe('1');
        });

        test('decrement should render -1', async () => {
          const decrement = renderedComponent.getByTestId('decrement');
          expect(decrement).not.toBeNull();

          await fireEvent.click(decrement);
          await tick();

          const count = renderedComponent.getByTestId('count');
          expect(count.innerHTML).toBe('-1');
        });

        test('reset should render 0', async () => {
          const reset = renderedComponent.getByTestId('reset');
          const increment = renderedComponent.getByTestId('increment');

          await fireEvent.click(increment);
          await tick();
          await fireEvent.click(reset);
          await tick();

          const count = renderedComponent.getByTestId('count');
          expect(count.innerHTML).toBe('0');
        });
      });
    });
  });
});
