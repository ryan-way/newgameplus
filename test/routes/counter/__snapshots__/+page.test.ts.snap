// Vitest Snapshot v1

exports[`Counter > after count has returned > updated component > should match snapshot 1`] = `undefined`;

exports[`Counter > after count has returned > updated component > should render counter as zero 1`] = `"0"`;

exports[`Counter > after count has returned > updated component > should render decrement button 1`] = `"Decrement"`;

exports[`Counter > after count has returned > updated component > should render increment button 1`] = `"Increment"`;

exports[`Counter > after count has returned > updated component > should render reset button 1`] = `"Reset"`;

exports[`Counter > when the component is rendered > should match snapshot 1`] = `undefined`;

exports[`Counter > when the component is rendered > should show the proper heading 1`] = `"Counter"`;
